### ISSUE TITLE
Linearize should be available from the OMEdit GUI

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#5626]
https://trac.openmodelica.org/OpenModelica/ticket/5626

**Comment by casella** [comment]
@adeas31, it's really time we implement this for good, every year I teach a course at TU Delft where I need it, and I need to go through the CLI thing, which is ugly. We need also some kind of choice between Modelica, Matlab, or Python output, which is set by the [--linearizationDumpLanguage](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-linearizationdumplanguage) flag.

Thanks!

**Comment by adeas31** [comment]
Looks like the [linearize](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html#linearize) API have same interface as simulate. So it makes sense to kind of integrate it in the existing simulation GUI interface.

I will investigate more how linearize is working and how we can integrate in a better way. Do you have any thing in mind how the GUI interface should be?

**Comment by casella** [comment]
Basically, linearize() does a regular simulation and then computes the A,B,C,D matrices of the linearize system at the end of the simulation, using the final state values to compute the jacobians"

Interface setup:
- Add an extra button after Simulate with Animation, call it Linearize, add a 2x2 matrix on the icon
- Also add a Linearize checkbox at the bottom of the General Tab of the Simulation window
- Add a drop-down menu right above C/C++ Compiler Flags (Optional), with the options of 
[`--linearizationDumpLanguage`](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-linearizationdumplanguage) flag. We can skip "none" which is not very useful. Default Modelica.
- The output is some generated file. The output text of the linearize function should be displayed in the Message window, it contains the exact location of the file

Test model: Modelica.Blocks.Continuous.Integrator

**Comment by casella** [comment]
@adeas31 can this make it for 1.26.0?

**Comment by casella** [comment]
Here is the updated specification:
- no extra button for Linearize, we just make it one option in Simulation Setup
- one extra tab after Output with title "Linearize"
- contents of the tab:
  - one checkbox (default not selected) "Linearize model at time = StopTime"
  - one drop-down menu with title "Target language for linearized model" the options of [`--linearizationDumpLanguage`](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-linearizationdumplanguage) except none: modelica, matlab, julia, python. Default: modelica

If the simulation and the linearization work fine, the simulation output contains something like:

LOG_STDOUT        | info    | Linear model is created at D:\\Temp\\OMEdit/linearized_model.m

This will appear in the messages window, I guess that will be enough, maybe it would be nice if the file name could be written in red.

**Comment by cbu-hw** [comment]
> maybe it would be nice if the file name could be written in red.

To me at least red text communicates "error/warning"... is that what you want to convey?

**Comment by adeas31** [comment]
> > maybe it would be nice if the file name could be written in red.
> 
> To me at least red text communicates "error/warning"... is that what you want to convey?

The message is send to standard out with severity set to `info`.  All info messages are shown in black text. If we want to do something different then we need to change it in runtime.

**Comment by casella** [comment]
Never mind that, not important.

**Comment by adeas31** [comment]
See PR #14811

Removed `none` option from `--linearizationDumpLanguage` flag. OMEdit has a checkbox to enable/disable linearization.

The output of linearization is not very good. It prints out the whole generated model and then prints its location and then a message,

```
The output format can be changed with the command line option --linearizationDumpLanguage.
The options are: --linearizationDumpLanguage=none, modelica, matlab, julia, python.
```

This message is not useful for GUI users. We should just print the location of the generated model.
Maybe someone who knows more about this can fix output messages in runtime.

**Comment by adeas31** [comment]
Also the model is generated in the working directory and not in `--outputPath`.

**Comment by phannebohm** [comment]
> Removed `none` option from `--linearizationDumpLanguage` flag. OMEdit has a checkbox to enable/disable linearization.

This was introduced to fix #12981. For large models we need a way to skip generating large C code if no linearization is needed. I guess you could also use `--maxSizeLinearization=0` for that (default is 1000 currently).

> The output of linearization is not very good. It prints out the whole generated model and then prints its location and then a message,
> 
> ```
> The output format can be changed with the command line option --linearizationDumpLanguage.
> The options are: --linearizationDumpLanguage=none, modelica, matlab, julia, python.
> ```
> 
> This message is not useful for GUI users. We should just print the location of the generated model. Maybe someone who knows more about this can fix output messages in runtime.

I can fix that tmrw.

**Comment by casella** [comment]
> This was introduced to fix [#12981](https://github.com/OpenModelica/OpenModelica/issues/12981). For large models we need a way to skip generating large C code if no linearization is needed.

Oops, I forgot about that, my bad.

>  I guess you could also use `--maxSizeLinearization=0` for that (default is 1000 currently).

That's one option.

**Comment by adeas31** [comment]
Does it make sense to have `--maxSizeLinearization` in OMEdit? If so, I can add it.

**Comment by casella** [comment]
Yes, absolutely, please add it.

**Comment by casella** [comment]
Further discussion during OSMC devmeeting. Decision on which language to use is taken at compile time, when the strings in the C code are actually generated. Can't change it at runtime, [-l](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-l) only specifies the time to carry out linearization _if the code to do so is present_.

So, it makes sense to re-instate the none option and actually make that default. No need to have the checkbox anymore, because that would be redundant, 

No need to edit `--maxSizeLinearization` in the GUI as well, because that too would be redundant
- in most cases, the generation of linearization code will be switched off, so no overhead at all
- if you switch it on, you know what your're doing. The current max size of 1000 is huge, it really makes no sense to generate ABCD matrices larger than that anyway, so there's no point making that editable.

Bottom line:

1.  the Linearize tab should just have a drop-down menu with all the options of [--linearizationDumpLanguage](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-linearizationdumplanguage), default `none`. Description text: Target language for linearized model
2. For the linearize API we keep backwards compatibility by setting it to `modelica` when it is set to the default `none`. This should be explained in the linearize() API function documentation.

**Comment by phannebohm** [comment]
> The current max size of 1000 is huge, it really makes no sense to generate ABCD matrices larger than that anyway, so there's no point making that editable.

The max size regards the (scalarized) sum `stateVars + inputVars + outputVars + algVars`. That is what makes the linearization code large/compilation slow, not the size of the matrices. The matrix size only affects the runtime performance.

**Comment by adeas31** [comment]
> Further discussion during OSMC devmeeting. Decision on which language to use is taken at compile time, when the strings in the C code are actually generated. Can't change it at runtime, [-l](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/simulationflags.html#simflag-l) only specifies the time to carry out linearization _if the code to do so is present_.
> 
> So, it makes sense to re-instate the none option and actually make that default. No need to have the checkbox anymore, because that would be redundant,
> 
> No need to edit `--maxSizeLinearization` in the GUI as well, because that too would be redundant
> 
> * in most cases, the generation of linearization code will be switched off, so no overhead at all
> * if you switch it on, you know what your're doing. The current max size of 1000 is huge, it really makes no sense to generate ABCD matrices larger than that anyway, so there's no point making that editable.
> 
> Bottom line:
> 
> 1. the Linearize tab should just have a drop-down menu with all the options of [--linearizationDumpLanguage](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-linearizationdumplanguage), default `none`. Description text: Target language for linearized model
> 2. For the linearize API we keep backwards compatibility by setting it to `modelica` when it is set to the default `none`. This should be explained in the linearize() API function documentation.

PR #14818 done all explained above.

@phannebohm the only thing left is to clean up the output messages of linearization (see https://github.com/OpenModelica/OpenModelica/issues/5626#issuecomment-3656387240) and make sure the file is created in `--outputPath` if it is specified.

**Comment by phannebohm** [comment]
Should be fixed with #14820. Please check and let me know if there is anything left to do.

**Comment by casella** [comment]
> The max size regards the (scalarized) sum `stateVars + inputVars + outputVars + algVars`. That is what makes the linearization code large/compilation slow, not the size of the matrices. 

OK. I mean, A has dimension stateVars*stateVars, B stateVars*inputVars, C outputVars*stateVars, D outputVars*inputVars, so the two quantities are correlated 😃

**Comment by casella** [comment]
Will check the nightly build tomorrow

**Comment by casella** [comment]
Looks nice. Thanks @adeas31!

**Comment by casella** [comment]
Please back-port to maintenance/v1.26

**Comment by adeas31** [comment]
Done with #14828
